be/src/storage/olap_server.cpp

```diff
@@ -395,6 +402,85 @@ void* StorageEngine::_pk_index_major_compaction_thread_callback(void* arg) {
     return nullptr;
 }
 
+void* StorageEngine::_local_pk_index_shard_data_gc_thread_callback(void* arg) {
+    if (!is_lake_persistent_index_dir_inited()) {
+        return nullptr;
+    }
+#ifdef GOOGLE_PROFILER
+    ProfilerRegisterThread();
+#endif
+    while (!_bg_worker_stopped.load(std::memory_order_consume)) {
+        SLEEP_IN_BG_WORKER(config::pindex_shard_data_gc_interval_seconds);
+        auto data_dir = get_persistent_index_store();
+        auto pk_path = data_dir->get_persistent_index_path();
+
+        LOG(INFO) << "start to gc local persistent index path:" << pk_path;
+
+        std::set<std::string> tablet_ids;
+        Status ret = fs::list_dirs_files(pk_path, &tablet_ids, nullptr);
+        if (!ret.ok()) {
+            LOG(WARNING) << "fail to walk dir. path=[" + pk_path << "] error[" << ret.to_string() << "]";
+            return nullptr;
+        }
+
+        auto lake_tablet_manager = ExecEnv::GetInstance()->lake_tablet_manager();
+        std::vector<int64_t> removed_meta_tablet_ids;
+        std::vector<int64_t> not_in_worker_tablet_ids;
+        std::vector<int64_t> removed_dir_tablet_ids;
+        WriteBatch wb;
+        for (const auto& tablet_id : tablet_ids) {
+            int64_t id = std::stoll(tablet_id);
+            if (!lake_tablet_manager->is_tablet_in_worker(id)) {
```

Comment:
How to resolve concurrent scenario between adding shard and cleaning index?

Reply:
Even if deleting the PK  in concurrent situation  by mistake， there should be no problem?

Reply:
This should be add lock